plexus-compiler-manager/src/main/java/org/codehaus/plexus/compiler/manager/DefaultCompilerManager.java

```diff
@@ -25,30 +25,51 @@
  */
 import javax.inject.Inject;
 import javax.inject.Named;
+import javax.inject.Provider;
 
 import java.util.Map;
 
 import org.codehaus.plexus.compiler.Compiler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * @author <a href="mailto:trygvis@inamo.no">Trygve Laugst&oslash;l</a>
  */
 @Named
 public class DefaultCompilerManager implements CompilerManager {
+    private static final String ERROR_MESSAGE = "Compiler '{}' could not be instantiated or injected properly. "
+            + "Running the build with -Dsisu.debug, looking for exceptions might help.";
+    private static final String ERROR_MESSAGE_DETAIL = "TypeNotPresentException caused by UnsupportedClassVersionError "
+            + "might indicate, that the compiler needs a more recent Java runtime. "
+            + "IllegalArgumentException in ClassReader.<init> might mean, that you need to upgrade Maven.";
+
     @Inject
-    private Map<String, Compiler> compilers;
+    private Map<String, Provider<Compiler>> compilers;
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
 
     // ----------------------------------------------------------------------
     // CompilerManager Implementation
     // ----------------------------------------------------------------------
 
     public Compiler getCompiler(String compilerId) throws NoSuchCompilerException {
-        Compiler compiler = compilers.get(compilerId);
+        // Provider<Class> is lazy -> presence of provider means compiler is present, but not yet constructed
+        Provider<Compiler> compilerProvider = compilers.get(compilerId);
 
-        if (compiler == null) {
+        if (compilerProvider == null) {
+            // Compiler could not be injected for some reason
+            log.error(ERROR_MESSAGE + " " + ERROR_MESSAGE_DETAIL, compilerId);
             throw new NoSuchCompilerException(compilerId);
         }
 
-        return compiler;
+        // Provider exists, but compiler was not created yet
+        try {
+            return compilerProvider.get();
+        } catch (Exception e) {
+            // DI could not construct compiler
+            log.error(ERROR_MESSAGE, compilerId);
+            throw new NoSuchCompilerException(compilerId);
```

Comment:
```suggestion
            throw new NoSuchCompilerException(compilerId, e);
```

Reply:
That was my first thought, too. But that constructor simply does not exist since 2005. Would you like me to add it? I can do that, if you think it makes sense.

Reply:
BTW, that code block is the one with the shorter error message. The longer one is in the block above, where there is no causing exception. I.e., there `-e` will not yield more information for the user. In order to see the exception from the block we are talking about here, I actually hade to add `if (true) throw new Exception("uh-oh")` during my manual tests.

Reply:
BTW, the thrown cause will be ignored by Maven Compiler:

https://github.com/apache/maven-compiler-plugin/blob/c30d0f6a2bf7210ab61996c49dcef0c0fbe207ea/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L680-L684

It just throws its own `MojoExecutionException` with the same message and compiler ID, i.e. whatever cause or other error message with more details I add, it has zero effect. Of course, we can open an issue in MCOMPILER and they can improve that, we cannot force all users to upgrade the compiler after the issue has been fixed.

```java
        try {
            compiler = compilerManager.getCompiler(compilerId);
        } catch (NoSuchCompilerException e) {
            throw new MojoExecutionException("No such compiler '" + e.getCompilerId() + "'.");
        }
```

See? This is the perfect example for my previous point: Let us clean in front of our own door and handle the issue as well as we can here, instead of relying on other components in the Maven ecosystem to always to the right thing.

Reply:
@laeubi: In ffa81def, I added an exception constructor taking and passing on a root cause for good measure, hoping that in the future if Maven Compiler handles it differently, it might be a little more useful than today. That does not make the error log obsolete, though.

Reply:
> See? This is the perfect example for my previous point: Let us clean in front of our own door and handle the issue as well as we can here, instead of relying on other components in the Maven ecosystem to always to the right thing.

No it is not a perfect example, it more shows what I mentioned before, `plexus-compiler` is an API / library .. you can't know how it is used and there is not only maven.

So to resolve this better for the user one needs two changes:

1. this one to include additional information at minimum the root cause (as done), maybe a different exception message (e.g. compiler with id was found but can't be loaded) so consumer sof plexus-compiler-api has everything we can offer here (e.g. that `-Dsisu.debug` actually works depends on the used implementation of the injection framework!)
2. another one in maven-compiler to include the causing exception in the `MojoExecutionException` and probably given more logging / instructions to the user how to handle this in the context of maven, for example it would be useful to log all jars the mojo has on its realm so a user can *see* what might be missing or is there ... and maybe giving a hint to `-Dsisu.debug` as we **know** maven is using that under the hood.

and then the logging becomes obsolete as you have already mentioned we can't even know *what* is wrong we can only see *that* something is wrong as the _why_ is dependent on the callers context!

Reply:
@laeubi, we agree to disagree then. FYI, I have just created [MCOMPILER-574](https://issues.apache.org/jira/browse/MCOMPILER-574) and mentioned you there.